### PR TITLE
Deprecate method parameters to DataFrame.join, DataFrame.merge.

### PR DESCRIPTION
**Pull request by bdice**
The `method` parameter to `DataFrame.join` and `DataFrame.merge` [isn't used internally](https://github.com/rapidsai/cudf/blob/e2098e56f0cb209b1d916ce617c04533444a056a/python/cudf/cudf/core/join/join.py#L90) after changes in #7454. This PR updates the docstrings and adds deprecation notices via `FutureWarning` as discussed in #9347.

The parameter is now deprecated in the public API. I removed all internal uses of the `method` parameter.

### CODE REVIEW

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -4605,6 +4604,8 @@ def join(
             column names when avoiding conflicts.
         sort : bool
             Set to True to ensure sorted ordering.
+        method :
+            This parameter is unused.
```
Comment (charlesbluca) [inline python/cudf/cudf/core/dataframe.py:4608]: Should we remove the `method` param from the `lhs.merge` call a couple lines later?

Comment (bdice) [reply python/cudf/cudf/core/dataframe.py:4608]: Yes, good catch. Resolved.

**Comment by bdice** [comment]
I just noticed the `method` argument is tested here: https://github.com/rapidsai/cudf/blob/6484e2ad573f3f331744f85d6cfd932476c09945/python/cudf/cudf/tests/test_joining.py#L105-L106

The tests should be updated to skip the `method` argument, then this will be ready for a final pass of review.

**Comment by bdice** [comment]
rerun tests

File: python/cudf/cudf/tests/test_joining.py [hunk 2]
```diff
@@ -136,8 +119,7 @@ def work_gdf(df):
     assert got.index.name is None
 
     assert list(expect.columns) == list(got.columns)
-    # test disabled until libgdf sort join gets updated with new api
-    if method == "hash":
+    if how in ["left", "inner", "right", "leftanti", "leftsemi"]:
```
Comment (bdice) [inline python/cudf/cudf/tests/test_joining.py:122]:
```suggestion
    if how in {"left", "inner", "right", "leftanti", "leftsemi"}:
```

**Comment by bdice** [comment]
@charlesbluca @brandon-b-miller This is now ready for review.

**Review by charlesbluca — APPROVED** [review]

**Review by brandon-b-miller — COMMENTED** [review]
[There doesn't seem to be a kwarg for this in pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html). Looking back through git it looks like this is a legacy parameter leftover from pygdf days. I do not see anywhere in the libcudf join header that consumes this either.

Should we deprecate or deprecate and remove this instead?

**Comment by bdice** [comment]
> [There doesn't seem to be a kwarg for this in pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html). Looking back through git it looks like this is a legacy parameter leftover from pygdf days. I do not see anywhere in the libcudf join header that consumes this either.
> 
> Should we deprecate or deprecate and remove this instead?

@shwina told me "I think we want to expose it, but perhaps with a warning that it does literally nothing." @shwina do you have additional thoughts on this? I vote for deprecation here.

**Comment by brandon-b-miller** [comment]
At some point we discussed what a python interface to `hash_join` would look like. Could be that we left it in planning to convert it over to be part of that API. But there were nuances around the memory behaving in puzzling ways to the user IIRC and we never built it out. I would still be weakly in favor of removing it until then though. 

Generally the pattern is that we write the function signature such that it either matches the pandas API exactly or matches exactly plus a superset of kwargs. Any parameters we don't support we detect and throw. this avoids users converting their code from pandas and possibly picking up quiet bugs from args/kwargs not being in the right spots.

**Comment by shwina** [comment]
This was my mistake @bdice. When we discussed this, my assumption was that Pandas exposed a `method=` parameter, and we needed to match that API. As (TIL) this is not the case, can we modify this PR to throw a deprecation warning instead? Let's plan to remove it entirely in the release after 21.12.

**Comment by bdice** [comment]
@shwina No problem! I'll make that addition. 👍

**Comment by shwina** [comment]
I created https://github.com/rapidsai/cudf/issues/9353 to track removal.

**Comment by shwina** [comment]
@gpucibot merge

**Pull request merged**